### ISSUE TITLE
issue with not in

### ISSUE THREAD

**Issue by lnu** [russellmcdonell/pyDMNrules#2]
Hello,
I'm trying to check if a value is not contained in a list. From my understanding, the rule should look like: `not aaa,bbb,ccc`.  
If I try a rule like that and it fails with:  
![image](https://user-images.githubusercontent.com/1829553/124544412-5cfb5c80-de27-11eb-935d-1f41feb45112.png)

If I try the same rule but without the `not`, which will be converted to an `in`, it works successfully.
I tried a `not in` directly in pysfeel and it works which means, I guess, that my rule is not correctly converted to the right sfeel syntax:
```
def test_notipysfeel(self):
        parser = pySFeel.SFeelParser()
        sfeelText = ' not("xxx" in("aaa","bbb","ccc"))'
        (status, retVal) = parser.sFeelParse(sfeelText)
        if "errors" in status:
            print("With errors:", status["errors"])
```

Did I miss something?

Thanks

**Comment by lnu** [comment]
I see a lot of `not(in` conversion and it seems wrong(from my understanding). So now I fixe my issue by doing this change:
![image](https://user-images.githubusercontent.com/1829553/124545700-cda37880-de29-11eb-9c39-d1d521b7f89e.png)

**Comment by russellmcdonell** [comment]
Try updating your copy of pySFeel - I fixed a bug in pySFeel recently - it
wasn't parsing 'in()' correctly for a list.
I'm not sure the bug is in pyDMNrules.

Russell McDonell

**Comment by lnu** [comment]
I tries with your HPV sample. If I do this:
![image](https://user-images.githubusercontent.com/1829553/125395117-09e75380-e3ab-11eb-95a5-ac3d18090c79.png)
I get the same error:`
![image](https://user-images.githubusercontent.com/1829553/125395136-14095200-e3ab-11eb-8bcb-936e8e5929d4.png)

but with this syntax it works:
![image](https://user-images.githubusercontent.com/1829553/125396624-14a2e800-e3ad-11eb-9057-8ec8d0a13992.png)

The fix is for this scenario

**Comment by russellmcdonell** [comment]
My interpretation of DMN is that this should fail.
The evaluation of each cell involves the input value, then an operator, then the cell contents, e.g.
FEEL(e in (t))=true

Your modification to my example would become
(HPV in (not in VU, V0, V1, V2))=true
That's not valid FEEL.

"in" is the default operator for a list.
"=" is the default operator for single value. e.g. 'O3' becomes
FEEL(Cyto-O = 'O3')=true

If there is an operator before the value, then that operator is used, e.g. < 24 becomes
FEEL(Patient.age < 24)=true

I do not believe there is an easy way of asserting "not in a list" in DMN. You can do it in FEEL, but not in a DMN.
You can do complex things like ' != null and (HPV not in ["VU", "V0", "V1", "V2"])'
which becomes
FEEL(HPV != null and (HPV not in ["VU", "V0", "V1", "V2"]))=true

However, a better strategy would be to have a preceding decision table that tests
HPV in VU, V0, V1, V2 and sets an output value of invalidHPV to false
with a second, default rule that sets invalidHPV true when HPV is not in VU, V0, V1, V2
Then use invalidHPV as an input in the second decision table and check that is it true.

**Comment by lnu** [comment]
it becomes:
![image](https://user-images.githubusercontent.com/1829553/125592537-4ac3e9dc-bc47-4865-b828-573f64a8d5d7.png)
So for you, we should not use this syntax?
Sorry for all the questions but I'm far from a specialist in DMN.

**Comment by russellmcdonell** [comment]
I wouldn't have done it that way, but that's not to say the syntax you found is invalid. And looking at the code, it does look as though I tried to support your original syntax. So this should stay as an issue until I determine what I was trying to do and whether or not I got it wrong.

However, the real power of DMN is the ability to have cascading decision tables. And for the output of one decision table to be the input of another. And to be able to control the flow through the decision tables, based upon outputs from previous decisions. The attached workbook is a trivial example of this.

Of course, if you do that newData changes from a dictionary to a list of dictionaries, where each of these dictionaries describes which decision table was run, and the outcome of that decision table. So, what you are probably looking for is no longer in newData{}, but is now in newData[-1]{}

[ExampleHPV_not.xlsx](https://github.com/russellmcdonell/pyDMNrules/files/6819821/ExampleHPV_not.xlsx)

**Comment by russellmcdonell** [comment]
I have released version 1.2.0 of pySFeel and version 1.2.0 of pyDMNrules. Both have bug fixes, including changes to 'in' and 'not'. I have included a set of unit tests so that I can do regression testing for any future releases. I've also uploaded a complex example (AN-SNAP rules (DMN).xlsx) to github. That complex example exposed quite a few bugs which I have now fixed.

**Comment by lnu** [comment]
Le met check this. Thank you!

**Comment by russellmcdonell** [comment]
There have been quite a few updates recently. Please let me know if this is still an issue. I would like to close this issue if this problem has gone away.

**Comment by russellmcdonell** [comment]
I have just released version 1.3.9 of pySFeel and version 1.3.18 of pyDMNrules.
I am closing this issue as this relates to a very old version, and bug does not seem to have reappeared recently.
